src/hlsBinaries.ts

```diff
@@ -79,364 +51,310 @@ class MissingToolError extends Error {
+        `haskell-language-server-wrapper${exeExt}`
+    );
+    const downloadedLegacyWrapper = path.join(
+        storagePath,
+        `haskell-language-server-wrapper${exeExt}`
+    );
+    let wrapper: string | undefined;
+    if (localWrapper) {
+        wrapper = localWrapper;
+    } else if (executableExists(downloadedWrapper)) {
+        wrapper = downloadedWrapper;
+    } else if (executableExists(downloadedLegacyWrapper)) {
+        wrapper = downloadedLegacyWrapper;
     }
-  }
-  // Not all users want to upgrade right away, in that case prompt
-  const updateBehaviour = workspace.getConfiguration('haskell').get('updateBehavior') as UpdateBehaviour;
 
-  if (updateBehaviour === 'never-check') {
-    logger.warn("As 'haskell.updateBehaviour' config option is set to 'never-check' " +
-      'we try to use the possibly obsolete cached release data');
-    return readCachedReleaseData();
-  }
-
-  try {
-    const releaseInfo = await httpsGetSilently(opts);
-    const releaseInfoParsed =
-      validate.parseAndValidate(releaseInfo, githubReleaseApiValidator).filter((x) => !x.prerelease) || null;
-
-    if (updateBehaviour === 'prompt') {
-      const cachedInfoParsed = await readCachedReleaseData();
-
-      if (
-        releaseInfoParsed !== null && releaseInfoParsed.length > 0 &&
-        (cachedInfoParsed === null || cachedInfoParsed.length === 0
-          || releaseInfoParsed[0].tag_name !== cachedInfoParsed[0].tag_name)
-      ) {
-        const promptMessage =
-          cachedInfoParsed === null
-            ? 'No version of the haskell-language-server is installed, would you like to install it now?'
-            : 'A new version of the haskell-language-server is available, would you like to upgrade now?';
-
-        const decision = await window.showInformationMessage(promptMessage, 'Download', 'Nevermind');
-        if (decision !== 'Download') {
-          // If not upgrade, bail and don't overwrite cached version information
-          return cachedInfoParsed;
+    const ghcup = path.join(storagePath, 'ghcup');
+    const updateBehaviour = workspace.getConfiguration('haskell').get('updateBehavior') as UpdateBehaviour;
+
+    // check if we need to update HLS
+    if (wrapper == null) {
+        // install new hls
+        if (updateBehaviour === 'never-check') {
+            throw new Error(
+                "No version of HLS installed or found and updateBehaviour set to 'never-check'" + 'giving up...'
+            );
+        } else if (updateBehaviour === 'prompt') {
+            const promptMessage =
+                'No version of the haskell-language-server is installed, would you like to install it now?';
+
+            const decision = await window.showInformationMessage(promptMessage, 'Download', 'Nevermind');
+            if (decision !== 'Download') {
+                throw new Error('No version of HLS installed or found and installation was denied' + 'giving up...');
```

Comment:
```suggestion
                throw new Error('No version of HLS installed or found and installation was denied, giving up...');
```

Again, space missing.